ocaml/typing/typeopt.ml

```diff
@@ -298,21 +339,25 @@ let rec value_kind env ~loc ~visited ~depth ~num_nodes_visited ty
         let visited = Numbers.Int.Set.add (get_id ty) visited in
         match decl.type_kind with
         | Type_variant (cstrs, rep) ->
-          value_kind_variant env ~loc ~visited ~depth ~num_nodes_visited cstrs rep
+          fallback_if_missing_cmi (num_nodes_visited, Pgenval) (fun () ->
+            value_kind_variant env ~loc ~visited ~depth ~num_nodes_visited cstrs
+              rep)
         | Type_record (labels, rep) ->
           let depth = depth + 1 in
-          value_kind_record env ~loc ~visited ~depth ~num_nodes_visited labels rep
+          fallback_if_missing_cmi (num_nodes_visited, Pgenval) (fun () ->
+            value_kind_record env ~loc ~visited ~depth ~num_nodes_visited labels
+              rep)
         | Type_abstract ->
           num_nodes_visited,
           value_kind_of_value_layout decl.type_layout
         | Type_open -> num_nodes_visited, Pgenval
-    with Not_found -> num_nodes_visited, Pgenval
+    with Not_found -> raise Missing_cmi_fallback
```

Comment:
The `Not_found` should come only from the call to `Env.find_type` some lines above, right? Then narrow the scope of the `try`/`with`.